minidns-dane-java7/src/main/java/de/measite/minidns/dane/DaneExtendedTrustManager.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2015 the original author or authors
+ *
+ * This software is licensed under the Apache License, Version 2.0,
+ * the GNU Lesser General Public License version 2 or later ("LGPL")
+ * and the WTFPL.
+ * You may choose either license to govern your use of this software only
+ * upon the condition that you accept all of the terms of either
+ * the Apache License 2.0, the LGPL 2.1+ or the WTFPL.
+ */
+package de.measite.minidns.dane;
+
+import de.measite.minidns.AbstractDNSClient;
+
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+import javax.net.ssl.X509ExtendedTrustManager;
+import javax.net.ssl.X509TrustManager;
+import java.net.Socket;
+import java.security.KeyManagementException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.util.logging.Logger;
+
+public class DaneExtendedTrustManager extends X509ExtendedTrustManager {
+    private final static Logger LOGGER = Logger.getLogger(DaneExtendedTrustManager.class.getName());
+
+    private final X509TrustManager base;
+    private final DaneVerifier verifier;
+
+    public static void inject() {
+        inject(new DaneExtendedTrustManager());
+    }
+
+    public static void inject(DaneExtendedTrustManager trustManager) {
```

Comment:
Is there any usage of having multiple trust managers? To my knowledge only the first entry in the array will be used. This is why you can define a "reference" trust manager in the constructor of DaneExtendedTrustManager.

This unfortunately does not work on Android, as the ExtendedTrustManager is a Java 7 feature (originally designed because of SNI) and we need the target URLs hostname to do the TLSA lookup. Maybe there is a comparable undocumented hack on Android, but I would not rely on undocumented features when implementing security...